sources/lib/run-time/llvm-posix-threads.c

```diff
@@ -0,0 +1,949 @@
+typedef struct mc {
+  pthread_mutex_t mutex;
+  pthread_cond_t  cond;
+} mc;
+
+typedef struct semaphore {
+  struct mc mc;
+  intptr_t count;
+  intptr_t max_count;
+} SEMAPHORE;
+
+typedef struct simple_lock {
+  struct mc mc;
+  pthread_t owner;
+  bool locked;
+} SIMPLELOCK;
+
+typedef struct recursive_lock {
+  struct mc mc;
+  pthread_t owner;
+  intptr_t recursion_count;
+} RECURSIVELOCK;
+
+typedef struct notification {
+  struct mc mc;
+} NOTIFICATION;
+
+
+/// Thread variables
+
+static pthread_mutex_t tlv_initializations_lock = PTHREAD_MUTEX_INITIALIZER;
+
+void primitive_register_thread_variable_initializer(dylan_value value, D function)
+{
+  pthread_mutex_lock(&tlv_initializations_lock);
+
+  // Add the value and function to %tlv-initializations
+  struct KLsimple_object_vectorGVKd *initializations
+    = (struct KLsimple_object_vectorGVKd *) Ptlv_initializations;
+  DSINT size = ((intptr_t) initializations->size) >> 2;
+
+  DSINT new_size = Ptlv_initializations_cursor + 2;
+  if (new_size > size) {
+    Ptlv_initializations
+      = primitive_grow_vector(Ptlv_initializations, 32);
+    initializations
+      = (struct KLsimple_object_vectorGVKd *) Ptlv_initializations;
+  }
+
+  initializations->vector_element[Ptlv_initializations_cursor++] = value;
+  initializations->vector_element[Ptlv_initializations_cursor++] = function;
+
+  pthread_mutex_unlock(&tlv_initializations_lock);
+}
+
+
+/// Thread operations
+
+static pthread_mutex_t thread_join_lock = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t thread_exit_event = PTHREAD_COND_INITIALIZER;
+
+static inline void thread_set_state(dylan_value t, uintptr_t state)
+{
+  struct KLthreadGYthreadsVdylan *thread
+    = (struct KLthreadGYthreadsVdylan *) t;
+  THREAD *rthread = (THREAD *) thread->handle1;
+  rthread->state = state;
+}
+
+static inline uintptr_t thread_get_state(dylan_value t)
+{
+  struct KLthreadGYthreadsVdylan *thread
+    = (struct KLthreadGYthreadsVdylan *) t;
+  THREAD *rthread = (THREAD *) thread->handle1;
+  return rthread->state;
+}
+
+void *get_current_thread_handle(void)
+{
+  return (void *) Pteb.teb_current_thread_handle;
+}
+
+static void *trampoline(void *arg)
+{
+  struct KLthreadGYthreadsVdylan *thread
+    = (struct KLthreadGYthreadsVdylan *) arg;
+
+  Pteb.teb_current_thread_handle = (D) pthread_self();
+
+  // Execute the thread function using the Dylan trampoline
+  dylan_value result = call_dylan_function(thread->handle2, 0);
+
+  // Drop references to the the <thread> object
+  Pteb.teb_current_thread = NULL;
+
+  // Mark the thread as completed
+  pthread_mutex_lock(&thread_join_lock);
+  thread_set_state(thread, thread_get_state(thread) | COMPLETED);
+  pthread_cond_broadcast(&thread_exit_event);
+  pthread_mutex_unlock(&thread_join_lock);
+
+  return result;
+}
+
+// primitive-make-thread
+dylan_value primitive_make_thread(dylan_value t, dylan_value n, dylan_value p, dylan_value f, DBOOL s)
+{
+  struct KLthreadGYthreadsVdylan *thread
```

Comment:
Come to think of it, I think we set the thread name elsewhere ... I'd have to look it up.